### PR TITLE
rfc(0011): mapping events to alerts

### PR DESCRIPTION
**Pull request by ncphillips**
Open for review until August 28th

### CODE REVIEW

**Comment by chrisdmacrae** [comment]
Two questions:

- how does the example plugin bind its default alerts to the CMS
- will an API for adding alerts dynamically during runtime still exist?

**Comment by ncphillips** [comment]
> * how does the example plugin bind its default alerts to the CMS

The CMS does that [when the API is registered](https://github.com/tinacms/rfcs/pull/12/files#diff-5281be4515b982e35e2e6931d7154e09R124-R134)

> * will an API for adding alerts dynamically during runtime still exist?

Yes this is a backwards compatible change.

**Comment by dwalkr** [comment]
> These Alerts can be replaced at runtime

The RFC doesn't appear to indicate how this would be done. Would a user need to re-instantiate the CMS?

I'm thinking about a multilingual TinaCMS setup, hypothetically using either (1) the browser's localization preferences, or (2) a toolbar widget / setting to switch the UI language

File: 0011-mapping-events-to-alerts.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  alerts,
+});
+```
+
+With this API, anytime the `something-cool` or the `something-bad` events are dispatched the corresponding function will be executed and
+the resulting Alert shown to the user.
+
+### Default Alert Maps in APIs
+
+Any API registered with the CMS could define default alerts for it's events.
```
Comment (dwalkr) [inline 0011-mapping-events-to-alerts.md:39]:
```suggestion
Any API registered with the CMS could define default alerts for its events.
```
so sorry

**Comment by dwalkr** [comment]
> But the proposed API does not expose that mapping

Exactly! I think this RFC should include a `replaceAlerts` method or something along those lines

**Comment by ncphillips** [comment]
Updated the implementation idea and added a "Updating dynamically" section. This stuff will live on the `Alerts` class not the `CMS` class. In hindsight that's obviously a better spot.

**Comment by dwalkr** [comment]
I'm going to leave some comments/questions about implementation. Not trying to get too deep in the weeds here as we can worry about specific implementation later, but I think it's relevant to some degree

File: 0011-mapping-events-to-alerts.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    // ...
+
+    this.alerts = new Alerts(this.events, options.alerts);
+  }
+
+  registerApi(name: string, api: any) {
+    // ...
+
+    if (this.api.alerts) {
+      this._eventsToAlerts = {
```
Comment (dwalkr) [inline 0011-mapping-events-to-alerts.md:131]: this could just call `this.alerts.setMap` now

File: 0011-mapping-events-to-alerts.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+type ToAlert = (
+  event: CMSEvent
+) => {
+  message: string;
+  level?: AlertLevel;
+  timeout?: number;
+};
+
+class TinaCMS extends CMS {
+  private _eventsToAlerts: EventsToAlerts = {};
```
Comment (dwalkr) [inline 0011-mapping-events-to-alerts.md:119]: `_eventsToAlerts`  wouldn't exist on `TinaCMS` now but be moved to the `Alerts` class, right?

File: 0011-mapping-events-to-alerts.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      this._eventsToAlerts = {
+        ...this._eventsToAlerts,
+        ...this.api.alerts,
+      };
+    }
+  }
+}
+
+class Alerts {
+  constructor(private events: EventBus, alerts = {}) {
```
Comment (dwalkr) [inline 0011-mapping-events-to-alerts.md:140]: I assume the constructor would call `this.setMap(alerts)`

File: 0011-mapping-events-to-alerts.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    this.events.subscribe('*', (event) => {
+      const toAlert = this._eventsToAlert[event.type];
+
+      if (toAlert) {
+        const { level, message, timeout } = toAlert(event);
+
+        this.alerts.add(level, message, timeout);
```
Comment (dwalkr) [inline 0011-mapping-events-to-alerts.md:147]:
```suggestion
        this.add(level, message, timeout);
```

**Comment by ncphillips** [comment]
Thank you! That was sloppy fixing them up now

**Comment by chrisdmacrae** [comment]
Based on the current iteration, this will work perfectly with the experimental i18n plugin, which is an indication to me it’s pretty close.

**Pull request merged**